### PR TITLE
Processor and Component for iframes

### LINKED ISSUE #281: Implement an iframe component
**Feature:** [Lazy load iframes processor](https://community.frontity.org/t/lazy-load-iframes-processor/1116)
**User Story**: Lazy load all iframes

- [x] Create the iframe component.
- [x] e2e tests.
- [x] Update docs.

Branch: `s0-f1-lazy-load-iframe`
[Sprint Card](https://github.com/orgs/frontity/projects/2#card-32042686)

### LINKED ISSUE #282: Implement an iframe processor
**Feature:** [Lazy load iframes processor](https://community.frontity.org/t/lazy-load-iframes-processor/1116)
**User Story**: Lazy load all iframes

- [x] Create the iframe processor.
- [x] Unit tests.
- [x] Update docs.

Branch: `s0-f1-lazy-load-iframe`
[Sprint Card](https://github.com/orgs/frontity/projects/2#card-32042686)

### PR DESCRIPTION
**Pull request by iamuchejude**
#### Description of what you did:

Create a processor and component for iframes.

#### My PR is a:

- 🔝 Improvement

#### Is the PR ready to be merged?

- ✅ Yes!

#### Feature Request

https://community.frontity.org/t/lazy-load-iframes-processor/1116

#### Docs Changes

https://app.gitbook.com/@frontity-tech/s/frontity/~/drafts/-M-TwT82Z6C5MTL5s2-v/api-reference-1/frontity-html2react

----

Fixes #281.
Fixes #282.

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+      "type": "node",
+      "request": "launch",
+      "name": "Launch Program",
+      "skipFiles": [
+        "<node_internals>/**"
+      ],
+      "program": "${file}"
+    }
+  ]
+}
```
Comment (luisherranz) [inline .vscode/launch.json]: Hey Jude, don't commit your `.vscode` folder. You can add it to the `.gitignore` file if you want :)

Comment (iamuchejude) [reply .vscode/launch.json]: Yea...I added it to my `.gitignore` in the script processor.

File: lerna-debug.log [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+0 silly argv   hoist: true,
+0 silly argv   lernaVersion: '3.15.0',
+0 silly argv   '$0': '../../../.npm/_npx/7456/bin/lerna'
+0 silly argv }
+1 notice cli v3.15.0
+2 verbose rootPath /Users/jude/Projects/frontity
+3 info versioning independent
+4 error JSONError: Unexpected end of JSON input while parsing near '' in packages/iframe/package.json
+4 error     at module.exports (/Users/jude/Projects/frontity/node_modules/parse-json/index.js:26:19)
+4 error     at parse (/Users/jude/Projects/frontity/node_modules/load-json-file/index.js:8:29)
```
Comment (luisherranz) [inline lerna-debug.log]: This file has to be deleted. It's just an error log from lerna.

File: examples/mars-theme-example/frontity.settings.js [hunk 3]
```diff
@@ -32,7 +32,7 @@ export default {
       name: "@frontity/wp-source",
       state: {
         source: {
-          api: "https://test.frontity.io/wp-json"
+          api: "http://frontity.local/wp-json"
```
Comment (luisherranz) [inline examples/mars-theme-example/frontity.settings.js]: Remember not to commit these changes.

File: packages/mars-theme/src/components/index.js [hunk 4]
```diff
@@ -41,6 +42,31 @@ const Theme = ({ state }) => {
           (data.isPostType && <Post />) ||
           (data.is404 && <Page404 />)}
       </Main>
+
+      <Iframe
```
Comment (luisherranz) [inline packages/mars-theme/src/components/index.js]: Same here, remember not to commit these changes.

Comment (iamuchejude) [reply packages/mars-theme/src/components/index.js]: Sure

File: examples/mars-theme-example/frontity.settings.js (hunk 3 above)
Comment (iamuchejude) [reply examples/mars-theme-example/frontity.settings.js]: Okay.

**Comment by luisherranz** [comment]
This is the table that represents the current approach to lazy loading in case you want to add it to the documentation:

| Intersection Observer | Native Lazy | Height > 0 | Output                |
|-----------------------|-------------|--------|-----------------------|
| true                  | true        | true   | Native Lazy Load             |
| true                  | true        | false  | Intersection Observer |
| true                  | false       | true   | Intersection Observer |
| true                  | false       | false  | Intersection Observer |
| false                 | true        | true   | (not possible)                     |
| false                 | true        | false  | (not possible)                     |
| false                 | false       | true   | Normal Load (eager)            |
| false                 | false       | false  | Normal Load (eager)            |

**Comment by luisherranz** [comment]
@DAreRodz could you please help @iamuchejude with the e2e tests of the iframe component?

By the way, I've removed the "HEADLESS" check because the headless version now uses Chrome 78. So now both IO and native lazy load should pass on headless as well 👍

**Comment by DAreRodz** [comment]
After having a call with @iamuchejude, we saw that some of the tests for the component using `IntersectionObserver` were failing because of a tiny error in the implementation. It's fixed now and they are passing, so that's great.

Regarding the native lazy-loading, the tests are passing in the normal browser, let's see if they pass in the headless browser as well. @iamuchejude, can you check that running `npm run e2e:run` in `frontity/e2e`?

**Comment by iamuchejude** [comment]
Okay, @DAreRodz. I'm on it.

**Comment by luisherranz** [comment]
> Regarding the native lazy-loading, the tests are passing in the normal browser, let's see if they pass in the headless browser as well.

I updated Cypress and headless is now using Chrome 78 so they should pass.

Great work guys 👍

**Review by luisherranz — APPROVED** [review]
Thanks @iamuchejude! I am going to merge this.

I've done some changes to your code thought, please take a look at my last commits:
- Added a changeset (remember to include one next time).
- Fixed a bug with classnames not having an space between them.
- Used ES5 syntax for the old-browser script.
- Remove screenshot from failed e2e test (don't commit those, although we should include this in the gitignore as well).

**Pull request merged**